src/main/java/com/google/devtools/build/lib/bazel/bzlmod/BazelModuleResolutionFunction.java

```diff
@@ -78,7 +78,7 @@ public SkyValue compute(SkyKey skyKey, Environment env)
     if (root == null) {
       return null;
     }
-    ImmutableMap<ModuleKey, Module> initialDepGraph = Discovery.run(env, root);
+    ImmutableMap<UnresolvedModuleKey, UnresolvedModule> initialDepGraph = Discovery.run(env, root);
```

Comment:
it's unclear to me why this initial dep graph is keyed by UnresolvedModuleKey, instead of ModuleKey. Is there a case where <foo, 1.2, 1> and <foo, 1.2, 2> correspond to different values in this map? From reading Discovery.java, it doesn't seem to be the case -- they both point to the result of ModuleFileFunction[<foo,1.2>].

So I suggest that we change "initialDepGraph" across the board to be keyed by ModuleKey, and rename "UnresolvedModuleKey" to something like "DepSpec" to clarify the fact that it's only ever used as the dependency specification in UnresolvedModule. (In fact UnresolvedModule also sounds a bit awkward to me. "Pre-selection module" would be better IMO; I'll keep thinking about this)

Reply:
> Can you look at the tests first and see if your thoughts still hold?

I had a brief look, and I don't think they change my opinion. Selection is basically "unpruned dep graph (ModuleKey->UnresolvedModule) -> resolved dep graph (ModuleKey->Module)". Just judging from this interface, there's nothing that requries a map whose key is UnresolvedModuleKey. If you need to build a UnresolvedModuleKey (really "DepSpec" is a better way to think about it) to ModuleKey mapping during Selection, then by all means do that; we shouldn't pollute the interface types with implementation needs.

> I think it's fine that before selection we are dealing with unresolved module keys.

My problem with that is it's conceptually wrong. The only thing that adding "max_compatibility_level" changes is that modules can ask for a special kind of "dep". It doesn't change the type of entities in the dependency graph, pre- or post-selection; it just means that we do some extra logic during selection. It's a tell that <foo, 1.2, 1> and <foo, 1.2, 2> will always point to the same object.

Reply:
I'll wait for all comments before making this change, but :+1:.